src/functional/ffront/foast_passes/type_deduction.py

```diff
@@ -107,8 +260,177 @@ def visit_FieldSymbol(
+    ) -> foast.TupleSymbol:
+        symtable = kwargs["symtable"]
+        if refine_type:
+            check_type_refinement(node, refine_type)
+            new_node = foast.TupleSymbol(id=node.id, type=refine_type, location=node.location)
+            symtable[new_node.id] = new_node
+            return new_node
+        return node
+
+    def visit_Subscript(self, node: foast.Subscript, **kwargs) -> foast.Subscript:
+        new_value = self.visit(node.value, **kwargs)
+        new_type = None
+        if kwargs.get("in_shift", False):
+            return foast.Subscript(
+                value=new_value, index=node.index, type=foast.OffsetType(), location=node.location
+            )
+        match new_value.type:
+            case foast.TupleType(types=types) | foast.FunctionType(
+                returns=foast.TupleType(types=types)
+            ):
+                new_type = types[node.index]
+            case _:
+                raise FieldOperatorTypeDeductionError.from_foast_node(
+                    new_value, msg="Could not deduce type of subscript expression!"
+                )
+
+        return foast.Subscript(
+            value=new_value, index=node.index, type=new_type, location=node.location
+        )
+
+    def visit_BinOp(self, node: foast.BinOp, **kwargs) -> foast.BinOp:
+        new_left = self.visit(node.left, **kwargs)
+        new_right = self.visit(node.right, **kwargs)
+        new_type = self._deduce_binop_type(
+            node.op, parent=node, left_type=new_left.type, right_type=new_right.type
+        )
+        return foast.BinOp(
+            op=node.op, left=new_left, right=new_right, location=node.location, type=new_type
+        )
+
+    def _deduce_binop_type(
+        self,
+        op: foast.BinaryOperator,
+        *,
+        parent: foast.BinOp,
+        left_type: foast.SymbolType,
+        right_type: foast.SymbolType,
+        **kwargs,
+    ) -> foast.SymbolType:
+        if op in [
+            foast.BinaryOperator.ADD,
+            foast.BinaryOperator.SUB,
+            foast.BinaryOperator.MULT,
+            foast.BinaryOperator.DIV,
+        ]:
+            return self._deduce_arithmetic_binop_type(
+                op, parent=parent, left_type=left_type, right_type=right_type, **kwargs
+            )
+        else:
+            return self._deduce_logical_binop_type(
+                op, parent=parent, left_type=left_type, right_type=right_type, **kwargs
+            )
+
+    def _binop_to_str(self, op: foast.BinaryOperator) -> str:
+        match op:
+            case foast.BinaryOperator.ADD:
+                return "+"
+            case foast.BinaryOperator.SUB:
+                return "-"
+            case foast.BinaryOperator.MULT:
+                return "*"
+            case foast.BinaryOperator.DIV:
+                return "/"
+            case foast.BinaryOperator.BIT_AND:
+                return "&"
+            case foast.BinaryOperator.BIT_OR:
+                return "|"
```

Comment:
This looks very much like the definition of the `foast.BinaryOperator.__str__()` function, I'd move it there.